File: 프로그래머스/3/42895. N으로 표현/N으로 표현.js
function solution(N, number) {
    var answer = 0;
    var a = Array.from(new Array(9), () => new Set());
    if(N==number) return 1;
    else {
        a.forEach((element, index) => {
            if (index !== 0) element.add(Number(String(N).repeat(index)));
        });
        for(var i=1 ; i<=8 ; ++i) {
            for(var j=1 ; j<i ; ++j) {
                for(var first of a[j]) {
                    for(var second of a[i-j]) {
                        a[i].add(first+second);
                        a[i].add(first-second);
                        a[i].add(first*second);
                        a[i].add(first/second);
                    }
                } 
            }
            if(a[i].has(number)) return i;
        }
        return -1;
    }
    return answer;
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N,number):
    if N == number:
        return 1
    # set은 중복제거
    dp = [set() for _ in range(9)] # dp[i] 는 N을 i번 이용해서 만들 수 있는 숫자들
    for i in range(1,9):
        dp[i].add(int(str(N) * i)) # 5 , 55, 555 만들려고 str*i 해준후 int로 변환
    
    for i in range(1,9):
        for j in range(1,i):
            for a in dp[j]:
                for b in dp[i-j]:
                    dp[i].add(a+b)
                    dp[i].add(a-b)
                    dp[i].add(a*b)
                    if b != 0:
                        dp[i].add(a//b)
        if number in dp[i]: # dp[i] 값이 number 되면 반환
            return i
    return -1